### PR TITLE
[SPARK-20886][CORE] HadoopMapReduceCommitProtocol to handle FileOutputCommitter.getWorkPath==null

### PR DESCRIPTION
**Pull request by steveloughran**
## What changes were proposed in this pull request?

Handles the situation where a `FileOutputCommitter.getWorkPath()` returns `null` by downgrading to the supplied `path` argument.

The existing code does an  `Option(workPath.toString).getOrElse(path)`, which triggers an NPE in the `toString()` operation if the workPath == null. The code apparently was meant to handle this (hence the getOrElse() clause, but as the NPE has already occurred at that point the else-clause never gets invoked.

## How was this patch tested?

Manually, with some later code review.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala
Comment (steveloughran) [inline core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala]: If this was intended to be resilient, it'd have to become `Option(workpath).getOrElse(path).toString()`, though that would actually hide the problem I'd managed to create: hand in a job committer to a task.

**Comment by jiangxb1987** [comment]
Could you also provide a test case to trigger the issue?

**Comment by steveloughran** [comment]
Not really. I thought about how I could do it, but essentially you do need to do things underneath the commit protocol, either in the Hadoop codebase (me) or in a test which somehow misconfigured things. It's most likely to surface in custom subclasses of `FileOutputCommitter`, so a test would probably consist of a run with a new committer set to always return null on `getWorkingPath`, verifying the changed assertion went from NPE to an IllegalArgException. 

That's a lot of complexity a test for what is essentially: making explicit a requirement of the code, namely, that `getWorkPath.toString()` requires `getWorkPath` to be non-null

I've included the before/after stack traces in the JIRA.

**Review by jiangxb1987 — APPROVED** [review]
LGTM, cc @cloud-fan

**Comment by steveloughran** [comment]
Is there anything else which needs to be one here, or is it matter of finding the right reviewer?

File: core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala]: if `workPath` is not null, will `workPath.toString` return null?

Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala]: I wonder the answer to this question ^ actually .. Wouldn't `Option(...).getOrElse(path)` be unnecessary?

Comment (steveloughran) [inline core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala]: {{workPath.toString()}} triggers an NPE, which was the reason for the stack trace.

Now, what's this code trying to do? find a directory to put stuff. If the committer is a subclass of `FileOutputFormat`,  then its getWorkPath method can be called, which *should* return a non-null path. 

If the requirement of the code is "get the workpath or return "path" if its null", maybe the code should really be

```scala
Option(f.getWorkPath).getOrElse(path).toString
```
That'd return the workPath if non null, falling back to the `path` variable, and then call `toString` on the returned object. I'm not sure off the top of my head if Scala type inference likes that though. It may be less elegant and more reliable to have

```scala
val workPath = f.getWorkPath
if (workPath != null) workPath.toString else path;
```
+maybe a bonus paranoid check for workPath being "".

I think that would actually get closer to the original goal of the code: always return a path, even of the committer doesn't have one

Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala]: Oh, maybe we could do this:

```scala
Option(f.getWorkPath).map(_.toString).getOrElse(path)
```

if I understood correctly.

**Comment by HyukjinKwon** [comment]
It looks good to me too.

**Comment by HyukjinKwon** [comment]
@steveloughran, mind fixing few words in the title and deacription accordingly? Let me leave this to @cloud-fan for a final look.

**Comment by HyukjinKwon** [comment]
OK. LGTM. I think this anyway should issue warnings:

```
WARN org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter: Output Path is null in setupJob()
WARN org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter: Output Path is null in commitJob()
```

and I assume we prefer working permissively rather than failing fast. At least, it looks that's what `FileOutputCommitter` allows even if it skips moving the output.

Please make a followup if I misunderstood and anything is wrong. Will merge this one in few days.

**Comment by steveloughran** [comment]
I believe this patch implements the original design goal: if a committer doesn't have a working path supplied by `getWorkingPath()` then it downgrades. 

It might be worthwhile doing a larger fault-injecting committer at some point.  have indirectly the ability to inject faults, but its a long way from the spark codebase, when really a thin shim is all that's needed. In particular, if a fault injecting ParquetOutputCommitter would test a big chunk of the codebase whose error handling codepaths are more likely to be tested in production than by jenkins. Are people interested (separate JIRA, obviously)

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by steveloughran** [comment]
thx

**Pull request closed without merging**
